src/query/functions/binary/node.go

```diff
@@ -0,0 +1,481 @@
+			return nil, errRightScalar
+		}
+
+		rVal := scalarR.Value()
+		// lhs is a series; use lhs metadata and series meta
+		return n.processSingleBlock(
+			lhs,
+			func(x float64) float64 {
+				return n.op.fn(x, rVal)
+			},
+		)
+	}
+
+	// both lhs and rhs are series
+	rIter, err := rhs.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	// NB(arnikola): this is a sanity check, as functions between
+	// two series missing vector matching should have previously
+	// errored out during the parsing step
+	if params.VectorMatching == nil {
+		return nil, errNoMatching
+	}
+
+	return n.processBothSeries(lIter, rIter)
+}
+
+func isComparison(op string) bool {
+	return op == EqType || op == NotEqType ||
+		op == GreaterType || op == LesserType ||
+		op == GreaterEqType || op == LesserEqType
+}
+
+// If returnBool is false and op is a comparison function,
+// the function should return the scalar value rather than 1 or 0
+func scalarFuncReturnBool(fn singleScalarFunc, op string, returnBool bool) singleScalarFunc {
+	if returnBool || !isComparison(op) {
+		return fn
+	}
+	return func(x float64) float64 {
+		take := fn(x) == 1
+		if take {
+			return x
+		}
+		// NB(arnikola): whereas prometheus does not include not matching
+		// points in the output, we here replace them with NaNs instead
+		return math.NaN()
+	}
+}
+
+func (n *binaryNode) processSingleBlock(
+	block block.Block,
+	fn singleScalarFunc,
+) (block.Block, error) {
+	returnBoolFn := scalarFuncReturnBool(fn, n.op.OperatorType, n.op.params.ReturnBool)
+
+	it, err := block.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	builder, err := n.controller.BlockBuilder(it.Meta(), it.SeriesMeta())
+	if err != nil {
+		return nil, err
+	}
+
+	if err := builder.AddCols(it.StepCount()); err != nil {
+		return nil, err
+	}
+
+	for index := 0; it.Next(); index++ {
+		step, err := it.Current()
+		if err != nil {
+			return nil, err
+		}
+
+		values := step.Values()
+		for _, value := range values {
+			builder.AppendValue(index, returnBoolFn(value))
+		}
+	}
+
+	return builder.Build(), nil
+}
+
+// If returnBool is false and op is a comparison function,
+// the function should return the scalar value of lhs if the comparison
+// is true; otherwise, it should return 1 if true, and 0 if false
+func funcReturnBool(fn mathFunc, op string, returnBool bool) mathFunc {
+	if returnBool || !isComparison(op) {
+		return fn
+	}
+	return func(left, right float64) float64 {
+		// NB (arnikola): safe to check that this equals 1 since this is
+		// only performed on comparison functions where viable values
+		// are only 1 or 0
+		take := fn(left, right) == 1
+		if take {
+			return left
+		}
+		// NB(arnikola): whereas prometheus does not include not matching
+		// points in the output, we here replace them with NaNs instead
+		return math.NaN()
+	}
+}
+
+func (n *binaryNode) processBothSeries(
+	lIter, rIter block.StepIter,
+) (block.Block, error) {
+	lMeta, rMeta := lIter.Meta(), rIter.Meta()
+
```

Comment:
Check to make sure bounds are equal

Reply:
it we're doing the check, should that be in process() ?